File: sdk/javascript/src/reliability/memoryQueue.ts
/**
 * MemoryQueue — the fast, in-process buffer `track()` writes into.
 * `put()` never performs I/O; for `overflowPolicy: "block"` it returns a
 * promise that resolves once room frees up (bounded by `blockTimeoutMs`)
 * rather than truly blocking Node's single-threaded event loop — other
 * work keeps running while it waits, matching how `track()` is already
 * async in this SDK.
 */

import type { OverflowPolicy, QueuedEvent } from "./types.js";

export interface MemoryQueueOptions {
  maxSize?: number;
  overflowPolicy?: OverflowPolicy;
  blockTimeoutMs?: number;
}

export class MemoryQueue {
  private readonly maxSize: number;
  private readonly overflowPolicy: OverflowPolicy;
  private readonly blockTimeoutMs: number;
  private items: QueuedEvent[] = [];
  private droppedCount = 0;
  private waiters: (() => void)[] = [];

  constructor(options: MemoryQueueOptions = {}) {
    this.maxSize = options.maxSize ?? 10_000;
    this.overflowPolicy = options.overflowPolicy ?? "drop_oldest";
    this.blockTimeoutMs = options.blockTimeoutMs ?? 1000;
    if (this.maxSize <= 0) throw new Error("maxSize must be positive");
  }

  /** Resolves True if the event was queued, False if dropped. */
  async put(event: QueuedEvent): Promise<boolean> {
    if (this.items.length < this.maxSize) {
      this.items.push(event);
      return true;
    }

    if (this.overflowPolicy === "drop_newest") {
      this.droppedCount += 1;
      return false;
    }

    if (this.overflowPolicy === "drop_oldest") {
      this.items.shift();
      this.items.push(event);
      this.droppedCount += 1;
      return true;
    }

    // block: wait for room, bounded by blockTimeoutMs.
    const gotRoom = await this.waitForRoom(this.blockTimeoutMs);
    if (!gotRoom) {
      this.droppedCount += 1;
      return false;
    }
    this.items.push(event);
    return true;
  }

  private waitForRoom(timeoutMs: number): Promise<boolean> {
    if (this.items.length < this.maxSize) return Promise.resolve(true);
    return new Promise((resolve) => {
      let settled = false;
      const timer = setTimeout(() => {
        if (settled) return;
        settled = true;
        resolve(false);
      }, timeoutMs);
      this.waiters.push(() => {
        if (settled) return;
        settled = true;
        clearTimeout(timer);
        resolve(true);
      });
    });
  }

  private notifyRoom(): void {
    if (this.items.length >= this.maxSize) return;
    const waiter = this.waiters.shift();
    if (waiter) waiter();
  }

  getBatch(maxItems: number): QueuedEvent[] {
    const batch = this.items.splice(0, maxItems);
    if (batch.length > 0) this.notifyRoom();
    return batch;
  }

  get size(): number {
    return this.items.length;
  }

  get dropped(): number {
    return this.droppedCount;
  }

  isEmpty(): boolean {
    return this.items.length === 0;
  }
}

File: sdk/python/costorah/reliability/memory_queue.py
"""
MemoryQueue — the fast, in-process buffer `track()` writes into.
`put()` never performs I/O and never blocks the caller for more than a
lock acquisition (microseconds), except when `overflow_policy="block"` is
configured — see the ticket's explicit "Overflow policy: Drop newest /
Drop oldest / Block. Configurable." requirement.
"""

from __future__ import annotations

import threading
from collections import deque

from costorah.reliability._types import OverflowPolicy, QueuedEvent


class MemoryQueue:
    def __init__(
        self,
        *,
        max_size: int = 10_000,
        overflow_policy: str = OverflowPolicy.DROP_OLDEST,
        block_timeout: float = 1.0,
    ) -> None:
        if max_size <= 0:
            raise ValueError("max_size must be positive")
        if overflow_policy not in OverflowPolicy.ALL:
            raise ValueError(f"overflow_policy must be one of {OverflowPolicy.ALL}")
        self._max_size = max_size
        self._overflow_policy = overflow_policy
        self._block_timeout = block_timeout
        self._items: deque[QueuedEvent] = deque()
        self._lock = threading.Lock()
        self._not_empty = threading.Condition(self._lock)
        self._not_full = threading.Condition(self._lock)
        self._dropped_count = 0

    def put(self, event: QueuedEvent) -> bool:
        """Returns True if the event was queued, False if it was dropped
        (only possible under drop_newest/drop_oldest at capacity)."""
        with self._lock:
            if len(self._items) < self._max_size:
                self._items.append(event)
                self._not_empty.notify()
                return True

            if self._overflow_policy == OverflowPolicy.DROP_NEWEST:
                self._dropped_count += 1
                return False

            if self._overflow_policy == OverflowPolicy.DROP_OLDEST:
                self._items.popleft()
                self._items.append(event)
                self._dropped_count += 1
                self._not_empty.notify()
                return True

            # BLOCK: wait for room, bounded by block_timeout so a caller
            # (and the whole application) can never hang indefinitely.
            deadline_ok = self._not_full.wait_for(
                lambda: len(self._items) < self._max_size, timeout=self._block_timeout
            )
            if not deadline_ok:
                self._dropped_count += 1
                return False
            self._items.append(event)
            self._not_empty.notify()
            return True

    def get_batch(self, max_items: int, *, timeout: float = 0.0) -> list[QueuedEvent]:
        """Drains up to `max_items`. With timeout=0 (default) this never
        blocks — used by the worker's poll loop. A positive timeout waits
        for at least one item, used by flush()/shutdown()."""
        with self._lock:
            if not self._items and timeout > 0:
                self._not_empty.wait(timeout=timeout)
            batch: list[QueuedEvent] = []
            while self._items and len(batch) < max_items:
                batch.append(self._items.popleft())
            if batch:
                self._not_full.notify_all()
            return batch

    def qsize(self) -> int:
        with self._lock:
            return len(self._items)

    @property
    def dropped_count(self) -> int:
        with self._lock:
            return self._dropped_count

    def is_empty(self) -> bool:
        with self._lock:
            return not self._items
